narwhals/_duckdb/utils.py

```diff
@@ -79,75 +80,68 @@ def evaluate_exprs(
+
+    if duckdb_dtype_id == "struct":
+        children = duckdb_dtype.children
         return dtypes.Struct(
             [
                 dtypes.Field(
-                    matchstruc_[i][0],
-                    native_to_narwhals_dtype(matchstruc_[i][1], version),
+                    name=child[0],
+                    dtype=native_to_narwhals_dtype(child[1], version=version),
                 )
-                for i in range(len(matchstruc_))
+                for child in children
             ]
         )
-    if match_ := re.match(r"(.*)\[\]$", duckdb_dtype):
-        return dtypes.List(native_to_narwhals_dtype(match_.group(1), version))
-    if match_ := re.match(r"(\w+)((?:\[\d+\])+)", duckdb_dtype):
-        duckdb_inner_type = match_.group(1)
-        duckdb_shape = match_.group(2)
-        shape = tuple(int(value) for value in re.findall(r"\[(\d+)\]", duckdb_shape))
-        return dtypes.Array(
-            inner=native_to_narwhals_dtype(duckdb_inner_type, version),
-            shape=shape,
-        )
-    if duckdb_dtype.startswith("DECIMAL("):
-        return dtypes.Decimal()
-    if duckdb_dtype == "TIME":
-        return dtypes.Time()
-    if duckdb_dtype == "BLOB":
-        return dtypes.Binary()
-    return dtypes.Unknown()  # pragma: no cover
+
+    if duckdb_dtype_id == "array":
+        child, size = duckdb_dtype.children
+        shape: list[int] = [size[1]]
+
+        while child[1].id == "array":
+            child, size = child[1].children
+            shape.insert(0, size[1])
+
+        inner = native_to_narwhals_dtype(child[1], version=version)
+        return dtypes.Array(inner=inner, shape=tuple(shape))
+
+    return _non_nested_native_to_narwhals_dtype(duckdb_dtype_id, version)
+
+
+@lru_cache(maxsize=16)
+def _non_nested_native_to_narwhals_dtype(duckdb_dtype_id: str, version: Version) -> DType:
+    dtypes = import_dtypes_module(version)
+    return {
+        "hugeint": dtypes.Int128(),
+        "bigint": dtypes.Int64(),
+        "integer": dtypes.Int32(),
+        "smallint": dtypes.Int16(),
+        "tinyint": dtypes.Int8(),
+        "uhugeint": dtypes.UInt128(),
+        "ubigint": dtypes.UInt64(),
+        "uinteger": dtypes.UInt32(),
+        "usmallint": dtypes.UInt16(),
+        "utinyint": dtypes.UInt8(),
+        "double": dtypes.Float64(),
+        "float": dtypes.Float32(),
+        "varchar": dtypes.String(),
+        "date": dtypes.Date(),
+        "timestamp": dtypes.Datetime(),
+        # TODO(marco): is UTC correct, or should we be getting the connection timezone?
+        # https://github.com/narwhals-dev/narwhals/issues/2165
+        "timestamp with time zone": dtypes.Datetime(time_zone="UTC"),
+        "boolean": dtypes.Boolean(),
+        "interval": dtypes.Duration(),
+        "decimal": dtypes.Decimal(),
+        "time": dtypes.Time(),
+        "blob": dtypes.Binary(),
+    }.get(duckdb_dtype_id, dtypes.Unknown())
```

Comment:
@MarcoGorelli sorry for the direct ping - I was wondering if should we create this dict for V1 and MAIN on top level on the module instead of for each call, or do it in another way? I didn't put too much thought into it 🤔

Reply:
I'd imagine this is fine

Reply:
Merging then 😉